feat: Redesign Tx flow modals

usame-algan opened the pull request:
## What it solves
Resolves #3082

## How this PR fixes it
Changes the layout/design of all Tx related modals

## Affected modals
- [x] SendFunds
- [x] ReviewSendFundsTx
- [x] SendCollectible
- [x] ReviewCollectible
- [x] ContractInteraction
- [x] ContractInteractionReview
- [x] SendCustomTx
- [x] ReviewCustomTx
- [x] ApproveTxModal
- [x] RejectTxModal
- [x] ConfirmTxModal
- [x] AddOwnerModal
- [x] RemoveOwnerModal
- [x] ReplaceOwnerModal
- [x] NewLimitModal
- [x] RemoveLimitModal

## Dependencies

- [x] Fix Typography for LimitModals once #3334 is merged
- [x] Rebase `epic/tx-details` to dev once #3297  is merged

## Issues found

- Modals are "jumpy" when deselecting execute transaction ([Screenshot](https://user-images.githubusercontent.com/5880855/151513647-207085a2-14a7-4549-b3be-1450a708f2b7.gif))
- Estimated Fees drop-down doesn't stay expanded if values are changed

## How to test it

1. Open the Safe App
2. Create Transactions to test each modal listed above
3. Compare with Figma Screens

katspaugh commented on src/routes/safe/components/Transactions/helpers/TxEstimatedFeesDetail/index.tsx:
```diff
@@ -0,0 +1,131 @@
+}: Props): ReactElement | null => {
+  const { nonce } = useSelector(currentSafe)
+  const threshold = useSelector(currentSafeThreshold) || 1
+  const defaultParameterStatus = isOffChainSignature && threshold > 1 ? 'ETH_HIDDEN' : 'ENABLED'
```
I've removed this in the approve modal refactoring, we'll need to merge carefully.

iamacook commented on src/routes/safe/components/Apps/components/ConfirmTxModal/DecodedTxDetail.tsx:
```diff
@@ -47,14 +45,6 @@ export const DecodedTxDetail = ({ hideDecodedTxData, onClose, decodedTxData: tx
-      />
-
-      <Hairline />
-
       <Container>{body}</Container>
```
You can the fragment.

usame-algan replied:
Good catch!

katspaugh commented:
There are a few inconsistencies wrt the Execute checkbox and showing/hiding certain parameters. All of which should go away once we merge #3369.

iamacook commented on src/routes/safe/components/Balances/SendModal/screens/ContractInteraction/Review/index.tsx:
```diff
@@ -96,11 +95,14 @@ const ContractInteractionReview = ({ onClose, onPrev, tx }: Props): React.ReactE
       onBack={onPrev}
     >
-      <ModalHeader onClose={onClose} subTitle="2 of 2" title="Contract interaction" />
+      <ModalHeader onClose={onClose} subTitle="Step 2 of 2" title="Contract interaction" />
```
Is this consistently named across all flows now?

iamacook replied:
For consistency, you could make a helper function to unify it all, e.g.:

```ts
const getStepTitle = (currentStep: number, totalSteps: number): string => `Step ${currentStep} of ${totalSteps}`
```

usame-algan replied:
Thats a clean solution I like that!

iamacook commented on src/routes/safe/components/Transactions/TxList/modals/ApproveTxModal.tsx:
```diff
@@ -361,25 +365,35 @@ export const ApproveTxModal = ({
                   </Paragraph>
 
-                  {oneConfirmationLeft && isExecution && !isCancelTx && <ExecuteCheckbox onChange={setShouldExecute} />}
+                  {txEstimationExecutionStatus === EstimationStatus.LOADING ? null : (
```
You need not use a ternary here.

iamacook commented on src/routes/safe/components/Transactions/TxList/modals/ApproveTxModal.tsx:
```diff
@@ -361,25 +365,35 @@ export const ApproveTxModal = ({
                 </Row>
               </Block>
 
               {txEstimationExecutionStatus === EstimationStatus.LOADING ? null : (
```
You can also remove the ternary here too.

iamacook commented on src/routes/safe/components/Transactions/TxList/modals/RejectTxModal.tsx:
```diff
@@ -120,7 +136,6 @@ export const RejectTxModal = ({ isOpen, onClose, gwTransaction }: Props): React.
 
               {txEstimationExecutionStatus === EstimationStatus.LOADING ? null : (
```
More ternaries that can be replaced.

iamacook commented on src/routes/safe/components/Transactions/helpers/TxEstimatedFeesDetail/index.tsx:
```diff
@@ -0,0 +1,131 @@
+  const lastQueuedTxNonce = useSelector(getLastTxNonce)
+
+  useEffect(() => {
+    if (Number.isNaN(safeNonceNumber)) return
+    if (safeNonceNumber === nonce) return
```
These can be added together with `||`.

usame-algan commented on src/components/ReviewInfoText/index.tsx:
```diff
@@ -57,12 +66,13 @@ export const ReviewInfoText = ({
       if (lastTxNonce !== undefined && safeNonceNumber === lastTxNonce + 1) return false
       return true
     }
+    // @FIXME: This can possibly be replace with useCanTxExecute
```
It is most likely outdated with #3369 so I will remove the comment

usame-algan commented on src/routes/safe/components/Settings/ManageOwners/ReplaceOwnerModal/screens/Review/index.tsx:
```diff
@@ -152,7 +152,7 @@ export const ReviewReplaceOwnerModal = ({
             </Row>
             <Row align="center" className={classes.info}>
-              <Paragraph color="primary" noMargin size="md" weight="bolder">
+              <Paragraph color="primary" noMargin size="xs" weight="bolder" style={{ letterSpacing: '1px' }}>
```
I added an `Overline` component as defined in the design system and reused it in the appropriate places.

usame-algan commented on src/components/ChainIndicator/index.tsx:
```diff
@@ -8,6 +8,7 @@ import { ChainId } from 'src/config/chain.d'
 interface Props {
   chainId: ChainId
   noLabel?: boolean
+  noCircle?: boolean
```
As far as I can see `noLabel` is never used so we could remove that prop but its used in more places with a circle than without so inverting to `showCircle` would require more changes.

iamacook replied:
Ok, I would remove `noLabel` and rename `noCircle` to `hideCircle` as it is semantically clearer.

iamacook reviewed (commented):
Looks amazing!

usame-algan commented on src/components/DecodeTxs/index.tsx:
```diff
@@ -140,70 +142,53 @@ export const getParameterElement = (parameter: DecodedDataBasicParameter, index:
-    <SingleTx decodedData={decodedData} onTxItemClick={onTxItemClick} />
-  )
+export const DecodeTxs = ({ txs, decodedData }: Props): ReactElement => {
+  return txs.length > 1 ? <MultiSendTx decodedData={decodedData} /> : <SingleTx decodedData={decodedData} />
```
Good catch, added a check for length 0

DiogoSoaress commented on src/routes/safe/components/Balances/SendModal/screens/ContractInteraction/SendCustomTx/index.tsx:
```diff
@@ -69,7 +69,7 @@ const SendCustomTx = ({ initialValues, isABI, onClose, onNext, switchMethod }: P
           Contract interaction
         </Paragraph>
-        <Paragraph className={classes.annotation}>1 of 2</Paragraph>
+        <Paragraph className={classes.annotation}>Step 1 of 2</Paragraph>
```
You could also use `getStepTitle(1,2)`

usame-algan replied:
Nice catch!

DiogoSoaress commented on src/routes/safe/components/Balances/SendModal/screens/ModalHeader/style.ts:
```diff
@@ -10,27 +10,28 @@ export const styles = createStyles({
-    padding: `0 ${md}`,
-    height: '20px',
+    fontSize: '12px',
+    padding: `5px ${sm}`,
```
Let's use the `variables` defined in the code as i mentioned above

DiogoSoaress replied:
I meant the `5px`

DiogoSoaress commented:
Toggling the "Execute transaction" makes the modal jumpy... Could we re-evaluate placing the check box above the accordeons (as before) ?
![jumpy modal](https://user-images.githubusercontent.com/32431609/151453615-2d7f8fb8-681a-4e35-bfa2-486141ea4bfe.gif)

iamacook commented:
> Could we re-evaluate placing the check box above the accordeons (as before)?

Why? We could use a [transation](https://v4.mui.com/components/transitions/#collapse).

usame-algan commented on src/components/ExecuteCheckbox/index.tsx:
```diff
@@ -12,15 +25,20 @@ const ExecuteCheckbox = ({ onChange }: ExecuteCheckboxProps): ReactElement => {
+        placement="top"
+        title="If you want to sign the transaction now but manually execute it later, click on the checkbox."
+      >
+        <span>
```
Our `Tooltip` component expects a wrapper in order to display the text content.

iamacook replied:
This is actually a inherent 'problem' with MUI.

usame-algan commented:
> Toggling the "Execute transaction" makes the modal jumpy

Compared to prod its more jumpy not only because the checkbox is now at the bottom but also because we have one more drop-down that disappears. I like the idea of moving the checkbox above the dropdowns to mitigate that effect.

usame-algan commented:
> Following this train of thought, the gas fee estimation accordion is not there whilst it is being initially calculated as well.

It would be nice to have a placeholder/skeleton in place with the same height while loading

iamacook commented:
The only real way of fixing the 'jumpy' issue is to fix the placement of the modal and ensure no elements above the checkbox unmount.

usame-algan commented:
@francovenica I am addressing your comments from #3380 here:

> Also, Right now the copy_to_clipboard / blockexplorer button are right at the limit of the modal, because of the chainName, this could become a problem if there are longer names. We should probably think of implementing ellipsis ( ... ) if this happens.

I would suggest we make the modal screens slightly larger to fix this. The address should always be displayed in a predictable way (full length or short hash with ellipsis in the middle).

> The replace owner modal looks different in the list of owners to be replaced. There is no design in figma for this so I assume is an unintentional change.

I removed the background in all owner related screens to keep consistency with the new design for adding an owner. Maybe we can use the red color in a more subtle way to go with the redesign. Could you give us input on this @liliiaorlenko ?

> The back button to remove a module doesn't work. It does nothing when you click it.
Same for the Change Threshold Modal.

This is fixed in #3369 

> Can we open the dropdowns after confirming edition?

The Advanced Parameters drop-down should still expand if the safe nonce changes but I think its a good idea to also expand the Estimated Fees drop-down if any of the parameters change to save a click.

katspaugh commented:
Let's just move the checkbox up.
Expanding a previously expanded dropdown would be nice but nothing critical.

francovenica commented:
@usame-algan 

- The color change in the "owner replace" modal I'd assume is something we should fix. I think it was just missed in the designs that should be this way.

- The copy_to_clipboard/explorer button is mostly something we should keep an eye on. No need to do anything now. It was mostly a warning for the designers that we might have to change things a little bit if there are longer names at some point
- The issues with the back button in the Upgrade safe modal and change policies are indeed fixed in the 3369, so it's all good.
- I agree with Ivan that the expand is a nice thing to have, but not a blocker

francovenica commented:
Ticket close and merged. Still leaving here that yes, indeed, the red color was fixed for owner replacement in the 3373. Thanks
![image](https://user-images.githubusercontent.com/4503659/151883686-54b650d1-6a70-4b9e-b4b1-643125a2d7a5.png)
